[WIP] Add type hints to datatypes

jdavcs opened the pull request:
There are numerous occurrences  of code like `dataset.metadata` in the code base, where `dataset` may refer to values of different types. This complicates reliably renaming the `model.DatasetInstance.metadata` property (see https://github.com/galaxyproject/galaxy/pull/13805#issuecomment-1138072044). Adding type hints is a step towards addressing this.

To do:
- [ ] Add type annotations (at least selectively) to the modules in `galaxy.datatypes`

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jdavcs committed:
Add type hints to datatypes.data

jdavcs committed:
Do not instantiate MarkdownFormatHelpers

No need: it's just static methods; also instead of passing references,
just import as needed.

jdavcs committed:
More type hints for datatypes.data

jdavcs committed:
Drop unnecessary check: data is not a string

jmchilton commented on lib/galaxy/datatypes/data.py:
```diff
@@ -34,17 +38,26 @@
+        HistoryDatasetAssociation,
+        HistoryDatasetCollectionAssociation,
+    )
+    from galaxy.tools import Tool
```
I'm -0 on all these forward references. We shouldn't be importing code we don't expect to be available in the packages. This includes GalaxyWebTransaction and Tool. The type checking will fail in those cases as we tighten up our mypy settings to disallow unknown imports.

-0 and not -1 because I will acknowledge fixes I would find palatable are challenging tasks of unwinding these dependencies or declaring protocols for how we consume app and trans in this package.

jdavcs replied:
My primary goal here is to add type annotations for dataset-like parameters (DatasetInstances + not DatasetInstances that are also "datasets" of another sort). So I can do without most other forward references. I'll drop GalaxyWebTransaction and Tool (I'll check the data package for imports to avoid whatever else I can). But yes, declaring protocols for those is a project of its own. If you don't mind I'll keep the dataset-related imports for now - I suppose we can replace them once we get to defining pydantic models for them (which I'd rather not tackle here).

jdavcs committed:
Limit forward references to data package

jdavcs committed:
Drop dataset parameter from datatype.get_visualizations()

jdavcs committed:
Add type annotations to set_meta() (wip)

jdavcs committed:
Check for None: TarFile.extractfile can return None

https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractfile

jdavcs committed:
Fix bug: don't drop args in call to parent

jdavcs committed:
Fix linting

jdavcs committed:
Fix mypy error for package tests (pysam)

jdavcs committed:
Add more type hints to datatypes/data.py

jdavcs committed:
Fix mypy by simplifying dict handling in anvio.py

jdavcs committed:
Add remaining type hints to data.py

jdavcs committed:
Add typing to definitions of dataset_content_needs_grooming

jdavcs committed:
Add typing to definitions of groom_dataset_content

jdavcs committed:
Add typing to definitions of init_meta

jdavcs committed:
Add typing to definitions of to_archive

jdavcs committed:
Add typing to definitions of display_as_markdown

jdavcs committed:
Add typing to definitions of _yield_user_file_content

jdavcs committed:
Raise NotImplementedError in base def; add typing for repair_methods

jdavcs committed:
Add typing to definitions of matches_any

jdavcs committed:
Add typing to definitions of get_visualizations

jdavcs committed:
Add typing to definitions of handle_dataset_as_image

jdavcs committed:
Add typing to definitions of line_dataprovider

jdavcs committed:
Add typing to definitions of regex_line_dataprovider

jdavcs committed:
Add typing to definitions of set_peek

jdavcs committed:
Remove commented out classes used by outdated versions of PlantTribes tools

Commented classes supported versions 1.0.0 - 1.0.2.
Current version is 1.0.4 from 2019.

jdavcs committed:
Refactor Ply class hierarchy to correct types

Text and Binary parent classes have different method signatures for
set_peek(). Ply is a base class whose subclasses may inherit from either
Text or Binary. Therefore, subclasses must define their own
implementation of set_peek (see Liskov substitution principle)

jdavcs committed:
Refactor Vtk class hierarchy (see Ply commit for details)

jdavcs committed:
Refactor Triples class hierarchy (see Ply commit for details)

jdavcs committed:
Add typing to definitions of get_test_fname

jdavcs committed:
Fix bug in get_file_peek + add type hints

On error, slicing was applied to string (default message), not list of lines read
from the file.

jdavcs committed:
Add typing to definitions of merge

jdavcs committed:
Fix bug: no `valid` method defined on DatatypeValidation

Must have been a typo: should be `validated`.

jdavcs committed:
Add typing to definitions of validate

jdavcs committed:
Add typing to definitions of split

jdavcs committed:
Add typing to definitions of get_mime

jdavcs committed:
Add typing to definitions of sniff

jdavcs committed:
Add missing return statements (fix mypy)

jdavcs committed:
Add property stubs for required attributes

Bug uncovered by mypy. Attrs must be defined in subclasses (as per
docstring comment).

jdavcs committed:
Add typing to definitions of "sniff_prefix"; add missing returns

jdavcs committed:
Add typing, fix arg naming in sniff_prefix in binary.py

Rename argument from sniff_prefix to file_prefix:
1. Consistent with all other definitions of the same method
2. FilePrefix is the correct type
3. Suboptimal naming (function name == argument name)

jdavcs committed:
Add typing to definitions of display_peek

jdavcs committed:
Fix mypy error in isa.py

jdavcs committed:
Fix bug in display_data in spaln.py and blast.py

The correct arg name is `dataset`, not `data`. The method body contains
a call to the same method in the parent class (defined in data.py),
passing `data` (i.e., dataset) as `data=data`, and that doesn't work,
because the parent method expexts `dataset`. Renaming the argument fixes
this error + the typing error that would've happened when
adding type hints.

jdavcs committed:
Rename parameter in display_data in data.py (see note)

Change from "data" to "dataset".
Use of this parameter is inconsistent across subclasses of Data when
they invoke the parent class method: some pass the HDA as "data", some
as "dataset"; as a result, values passed as "dataset" were lost.

I chose "dataset" becuase it's more consistent with how we typically refer to an
HDA argument in the code base. It may have been more accurate to use
"hda" instead, but there are numerous similar cases that would need to
be renamed following this pattern for consistency (even across the
datatypes directory).

jdavcs committed:
Add typing to definitions of display_data; fix mypy errors

jdavcs committed:
Add typing to definitions of set_meta; fix mypy errors

jdavcs committed:
Remove return statements from set_meta in genetics.py

There should be no return statements in those methods.

jdavcs committed:
Correct file permissions of goldenpath.py (remove exec)

jdavcs committed:
Add typing to definitions of generate_primary_file

jdavcs committed:
Fix bug in anvio.py: check `if dataset` before accessing its attribute
